ouroboros-network/test/Test/Ouroboros/Network/Testnet/Simulation/Node.hs

```diff
@@ -0,0 +1,446 @@
+                   BearerInfo (..))
+
+import           Test.Ouroboros.Network.Diffusion.Node.NodeKernel (NtNAddr, randomBlockGenerationArgs, BlockGeneratorArgs)
+import           Test.Ouroboros.Network.PeerSelection.RootPeersDNS (DNSTimeout,
+                   DNSLookupDelay)
+import qualified Test.Ouroboros.Network.Diffusion.Node.NodeKernel    as Node
+import qualified Test.Ouroboros.Network.Diffusion.Node.MiniProtocols as Node
+import qualified Test.Ouroboros.Network.Diffusion.Node               as Node
+
+import           Test.QuickCheck
+                   (Arbitrary(arbitrary),
+                     Gen,
+                     choose,
+                     chooseInt,
+                     sized,
+                     vectorOf,
+                     suchThat,
+                     oneof,
+                     frequency,
+                     sublistOf
+                   )
+
+
+-- | Diffusion Simulator Arguments
+--
+-- Contains all necessary randomly generated values needed to run diffusion in
+-- simulation.
+--
+data SimArgs =
+  SimArgs
+    { saSlot :: DiffTime
+      -- ^ 'randomBlockGenerationArgs' slot duration argument
+    , saSeed :: StdGen
+      -- ^ 'randomBlockGenerationArgs' seed argument
+    , saQuota :: Int
+      -- ^ 'randomBlockGenerationArgs' quota value
+    , saMbTime :: Maybe DiffTime
+      -- ^ 'LimitsAndTimeouts' argument
+    , saRelays :: [RelayAccessPoint]
+      -- ^ 'Interfaces' relays auxiliary value
+    , saRng :: StdGen
+      -- ^ 'Interfaces' 'iRng' value
+    , saDomainMap :: Map Domain [IP]
+      -- ^ 'Interfaces' 'iDomainMap' value
+    , saAddr :: NtNAddr
+      -- ^ 'Arguments' 'aIPAddress' value
+    , saLocalRootPeers :: [(Int, Map RelayAccessPoint PeerAdvertise)]
+      -- ^ 'Arguments' 'LocalRootPeers' values
+    , saLocalSelectionTargets :: PeerSelectionTargets
+      -- ^ 'Arguments' 'aLocalSelectionTargets' value
+    , saDNSTimeoutSctipt :: Script DNSTimeout
+      -- ^ 'Arguments' 'aDNSTimeoutScript' value
+    , saDNSLookupDelayScript :: Script DNSLookupDelay
+      -- ^ 'Arguments' 'aDNSLookupDelayScript' value
+    }
+    deriving (Show)
+
+data Commands = JoinNetwork DiffTime
+               | Kill DiffTime
+               | Reconfigure DiffTime
+                             [(Int, Map RelayAccessPoint PeerAdvertise)]
+  deriving (Show, Eq)
+
+-- | Generate DNS table
+genDomainMap :: [RelayAccessPoint] -> Gen (Map Domain [IP])
+genDomainMap raps = do
+  let domains = [ d | RelayAccessDomain d _ <- raps ]
+  m <- mapM (\d -> do
+    size <- chooseInt (1, 5)
+    ips <- vectorOf size genIP
+    return (d, ips)) domains
+
+  return (Map.fromList m)
+
+  where
+    genIP :: Gen IP
+    genIP =
+      let genIPv4 = IPv4 . toIPv4 <$> replicateM 4 (choose (0,255))
+          genIPv6 = IPv6 . toIPv6 <$> replicateM 8 (choose (0,0xffff))
+       in oneof [genIPv4, genIPv6]
+
+genCommands :: [(Int, Map RelayAccessPoint PeerAdvertise)] -> Gen [Commands]
+genCommands raps = sized $ \size -> do
```

Comment:
This will only generate a list which starts with `JoinNetwork` and continues with one of `Reconfigure` or `Kill`, and a kill can be followed by a single `JoinNetwork`.  It would be good to cover `JoinNetwork` after being `Kill`-ed in a more general way.

If you want to look at inspiration how this can be done more generally, take a look at
 [Test.Ouroboros.Network.Server2](https://github.com/input-output-hk/ouroboros-network/blob/master/ouroboros-network-framework/test/Test/Ouroboros/Network/Server2.hs?plain=1#L1235) - which is a statefull generator.  Here we'd need to know if a node is running or not. (I don't meant it must be done this way).
 
We don't include here attenuations which kill the node at specific times, e.g. an accept error, but in the future we could extend this generator to also generate attenuation as well.  Right now, with decoupled node state and its attenuation, we will need to be lucky to have `Kill` followed by `JoinNetwork` to restart after an attenuation error.  For this reason it's better to have slightly more nodes.

It would be nice to have some parameter which allows us to configure how often a node is restarted and have nodes which are restarted often and some others that are restarted more rarely (and thus are more stable).

Reply:
There's a different way I could do this that is by generating an arbitrary list of commands and then fixing it up. We shouldn't have 2 consecutive `Kill` or 2 consecutive `JoinNetwork`, and after a `Kill` the only that makes sense to follow it is `JoinNetwork`. So something like this:

```haskell
fixupCommands :: [Command] -> [Command]
fixupCommands [] = []
fixupCommands (h:t) = go h t
  where
    go :: Command -> [Command] -> [Command]
    go _ [] = []
    go prev (cmd:cmds) =
      case (prev, cmd) of
        (JoinNetwork _, JoinNetwork dt) ->
          JoinNetwork dt : go cmd cmds
        (Kill _, Kill dt) ->
          Kill dt : go cmd cmds
        (Kill dt, Reconfigure _ _) ->
          Kill dt : go prev cmds
        (Reconfigure dt lrp, JoinNetwork _) ->
          Reconfigure dt lrp : go prev cmds
        _ -> prev : go cmd cmds

```

Reply:
^ it does not guarantee that `Kill` or `Reconfigure` runs when a node is running.

Reply:
Doesn't it? I'd say otherwise but might be overlooking something, can you point me to it?

Reply:
e.g. `[Kill]` or `[Kill, Reconfigure, Kill, Reconfigure]` are both fixpoints of `fixupCommands`; I think you need to track state if a node is running or not. There might be an arbitrary long strides of `Reconfigure` until a node is `Kill`-ed when it must be restarted.

Reply:
Oh right! I think singleton lists are the only thing I missed, I should add a clause:

```haskell
    go _ [] = []
    go (JoinNetwork dt) [] = [JoinNetwork dt]
    ...
```

I think this ought to fix the problem. For other singleton lists I will ignore them and return `[]`. I dont think I need to keep track of state for `fixupCommands`, I can just assume the node is running at the current recursion point and decide what to do with whatever 2 consecutive commands are there next. So for example I can not have 2 `Kills` consecutively and also a `Kill` after a `Reconfigure` doesn't make much sense to. 

> There might be an arbitrary long strides of Reconfigure until a node is Kill-ed when it must be restarted.

Are you saying after a `Reconfigure` I need to restart the node? Because if that wouldn't really be necessary, the node handles that by itself.

Reply:
I am convinced. 

> Are you saying after a Reconfigure I need to restart the node? Because if that wouldn't really be necessary, the node handles that by itself

no.

Reply:
Addressed